test/run_e2e_test.py

```diff
@@ -0,0 +1,195 @@
+
+    def __init__(self, future: Future, expected_lifecycle_nodes: List):
+        super().__init__('statisticsListener')
+
+        self.sub = self.create_subscription(MetricsMessage,
+                                            'system_metrics',
+                                            self.listener_callback,
+                                            1)
+        self.future = future
+        self.received_all_published_stats = False
+        self.expected_lifecycle_nodes = expected_lifecycle_nodes
+
+    def listener_callback(self, msg) -> None:
+        """
+        Handle published MetricsMessages.
+
+        :param msg: received message
+        :return: None
+        """
+        node_name = '/' + msg.measurement_source_name
+
+        if node_name in self.expected_lifecycle_nodes:
+            self.expected_lifecycle_nodes.remove(node_name)
+
+        if not self.expected_lifecycle_nodes:
+            self.received_all_published_stats = True
+            self.future.set_result(self.received_all_published_stats)
+
+
+def print_test(message: str) -> None:
+    """
+    Print a formatted string for easier parsing.
+
+    :param message: string to be printed
+    :return:
+    """
+    print('========================')
+    print(message)
+    print('========================')
+
+
+def check_for_expected_nodes(args=None) -> None:
+    """
+    Check that all expected nodes can be found.
+
+    :param args:
+    :return:
+    """
+    expected_nodes = ['/listener', '/talker'] + EXPECTED_LIFECYCLE_NODES
+
+    rclpy.init(args=args)
+    node = rclpy.create_node('check_for_expected_nodes_test')
+    actual_nodes = node.get_node_names_and_namespaces()
+    node.destroy_node()
+    rclpy.shutdown()
+
+    expected_count = len(expected_nodes)
+    actual_count = 0
+    for item in actual_nodes:
+        found_node = '/' + item[0]
+        if found_node in expected_nodes:
+            actual_count += 1
+            expected_nodes.remove(found_node)
+
+    if expected_count != actual_count:
+        raise Exception('Failed to enumerate expected nodes. Found #' + actual_count + ': ', actual_nodes)
+    else:
+        print_test('check_for_expected_nodes success')
+
+
+def check_lifecycle_node_enumeration() -> None:
+    """
+    Check that all lifecycle nodes exist.
+
+    This requires the ros2lifecycle dependency.
+    :return:
+    """
+    stream = os.popen('ros2 lifecycle nodes')
+    output = stream.read().splitlines()
+    if output != EXPECTED_LIFECYCLE_NODES:
+        raise Exception('check_lifecycle_node_enumeration failed: ' + str(output))
+    else:
+        print_test('check_lifecycle_node_enumeration success')
+
+
+def check_lifecycle_node_state() -> None:
+    """
+    Check that each lifecycle node is active.
+
+    This requires the ros2lifecycle dependency.
+    :return:
+    """
+    for lifecycle_node in EXPECTED_LIFECYCLE_NODES:
+        stream = os.popen('ros2 lifecycle get ' + str(lifecycle_node))
```

Comment:
I think it would be better to use the [`subprocess`](https://docs.python.org/3.8/library/subprocess.html) module to handle stdout/err streams and result codes better.

Reply:
I'll use ```subprocess.check_output```